org.eclipse.paho.client.mqttv3/src/main/java/org/eclipse/paho/client/mqttv3/internal/ClientState.java

```diff
@@ -698,49 +676,34 @@ public MqttToken checkForActivity(IMqttActionListener pingCallback) throws MqttE
-                if ((pingOutstanding == 0 && (time - lastInboundActivity >= keepAlive - delta)) ||
-                    (time - lastOutboundActivity >= keepAlive - delta)) {
-
-                    //@TRACE 620=ping needed. keepAlive={0} lastOutboundActivity={1} lastInboundActivity={2}                                                                                                              
-                    log.fine(CLASS_NAME,methodName,"620", new Object[]{new Long(this.keepAlive),new Long(lastOutboundActivity),new Long(lastInboundActivity)});
-
-                    // pingOutstanding++;  // it will be set after the ping has been written on the wire                                                                                                             
-                    // lastPing = time;    // it will be set after the ping has been written on the wire                                                                                                             
-                    token = new MqttToken(clientComms.getClient().getClientId());
-                    if(pingCallback != null){
-                    	token.setActionCallback(pingCallback);
-                    }
-                    tokenStore.saveToken(token, pingCommand);
-                    pendingFlows.insertElementAt(pingCommand, 0);
-
-                    nextPingTime = getKeepAlive();
-
-                    //Wake sender thread since it may be in wait state (in ClientState.get())                                                                                                                             
-                    notifyQueueLock();
-                }
-                else {
-                    log.fine(CLASS_NAME, methodName, "634", null);
-                    nextPingTime = Math.max(1, getKeepAlive() - (time - lastOutboundActivity));
-                }
+                // Always send a ping if we get here
+				//@TRACE 620=ping needed. keepAlive={0} lastOutboundActivity={1} lastInboundActivity={2}
+				log.fine(CLASS_NAME,methodName,"620", new Object[]{this.keepAlive, lastOutboundActivity, lastInboundActivity});
+
+				// pingOutstanding++;  // it will be set after the ping has been written on the wire
+				// lastPing = time;    // it will be set after the ping has been written on the wire
+				token = new MqttToken(clientComms.getClient().getClientId());
+				if(pingCallback != null){
+					token.setActionCallback(pingCallback);
+				}
+				tokenStore.saveToken(token, pingCommand);
+				pendingFlows.insertElementAt(pingCommand, 0);
+
+				//Wake sender thread since it may be in wait state (in ClientState.get())
+				notifyQueueLock();
+
+				// Schedule the next ping at half the keep alive interval to ensure we don't get disconnected
+				nextPingTime = getKeepAlive() / 2;
```

Comment:
This all seems reasonable.

Reply:
hmm, looks like the file is a mix of tabs and spaces.